Refactor implicit evaporator as harmonic barrier

kkritikak opened the pull request:
This PR reimplements the implicit evaporator as a more general harmonic barrier. The refactored code uses a template to reduce code duplication between geometries. This will allow, for example, adding a cylinder geometry in future.

mphoward commented on src/CMakeLists.txt:
```diff
@@ -6,11 +6,16 @@ set(_${COMPONENT_NAME}_sources
     module.cc
+    MovingHarmonicPotential.cc
     ParabolicFlow.cc
+    PlanarMovingHarmonicBarrier.cc
```
I'm really sorry... looking at these, I realized I made a mess of these names. Could you please rename one more time as:
* `HarmonicBarrier`
* `PlanarHarmonicBarrier`
* `SphericalHarmonicBarrier`

mphoward commented on src/MovingHarmonicPotential.cc:
```diff
@@ -3,50 +3,45 @@
-    m_pdata->getNumTypesChangeSignal()
-        .disconnect<ImplicitEvaporator, &ImplicitEvaporator::reallocateParams>(this);
-    }
+MovingHarmonicPotential::~MovingHarmonicPotential() { };
```
`;` not needed after function definition.
```suggestion
MovingHarmonicPotential::~MovingHarmonicPotential() { }
```

mphoward commented on src/MovingHarmonicPotential.cc:
```diff
@@ -3,50 +3,45 @@
     PDataFlags flags = m_pdata->getFlags();
     if (!m_has_warned
-        && (flags[pdata_flag::pressure_tensor] || flags[pdata_flag::isotropic_virial]))
+        && flags[pdata_flag::pressure_tensor])
```
This is correct! See here:
https://github.com/glotzerlab/hoomd-blue/blob/58b7ba503cadd90cc1ed65f40208a8d2de5b7b63/hoomd/ParticleData.h#L68-L78

mphoward commented on src/MovingHarmonicPotential.cc:
```diff
@@ -58,15 +53,17 @@ namespace detail
-    py::class_<ImplicitEvaporator, std::shared_ptr<ImplicitEvaporator>>(m,
-                                                                        "ImplicitEvaporator",
-                                                                        py::base<ForceCompute>())
+    py::class_<MovingHarmonicPotential, std::shared_ptr<MovingHarmonicPotential>>(m,
+                                                                                  "MovingHarmonicPotential",
+                                                                                  py::base<ForceCompute>())
```
We recently discovered that this way of describing the inheritance of the class leads to very strange (incorrect) behavior in Python, due to some change in pybind11. I think it should be like this instead:
```suggestion
    py::class_<HarmonicBarrier, std::shared_ptr<HarmonicBarrier>, ForceCompute>(m,
                                                                                  "HarmonicBarrier")
```
You will want to propagate this everywhere.

mphoward commented on src/MovingHarmonicPotential.cc:
```diff
@@ -58,15 +53,17 @@ namespace detail
+                                                                                  py::base<ForceCompute>())
         .def(py::init<std::shared_ptr<SystemDefinition>, std::shared_ptr<Variant>>())
-        .def("setParams", &ImplicitEvaporator::setParams);
+        .def("setParams", &MovingHarmonicPotential::setParams);
```
To work with the per-type Python interface, I think this is going to need to be modified to accept a data structure for this class that contains all the parameters. That data structure is also going to need a constructor that creates it from a Python dictionary (and I think one that can go the other way too).

Take a look at the methods here for what this class needs as an example:

https://github.com/glotzerlab/hoomd-blue/blob/58b7ba503cadd90cc1ed65f40208a8d2de5b7b63/hoomd/md/PotentialExternal.h#L185

and here for how the parameter class gets implemented:

https://github.com/glotzerlab/hoomd-blue/blob/58b7ba503cadd90cc1ed65f40208a8d2de5b7b63/hoomd/md/EvaluatorExternalMagneticField.h#L48-L76

mphoward commented on src/MovingHarmonicPotentialGPU.cc:
```diff
@@ -16,9 +19,9 @@ namespace azplugins
+                                                       std::shared_ptr<Variant> interf)
+    : MovingHarmonicPotential(sysdef, interf)
     {
     m_tuner.reset(new Autotuner(32, 1024, 32, 5, 100000, "implicit_evap", m_exec_conf));
```
The way that autotuners get created and assigned has changed. See here for an example:

https://github.com/glotzerlab/hoomd-blue/blob/58b7ba503cadd90cc1ed65f40208a8d2de5b7b63/hoomd/md/PotentialExternalGPU.h#L49-L52

The member variable is also going to look different:

https://github.com/glotzerlab/hoomd-blue/blob/58b7ba503cadd90cc1ed65f40208a8d2de5b7b63/hoomd/md/PotentialExternalGPU.h#L39

mphoward commented on src/MovingHarmonicPotentialGPU.h:
```diff
@@ -41,7 +44,7 @@ class PYBIND11_EXPORT ImplicitEvaporatorGPU : public ImplicitEvaporator
      */
     virtual void setAutotunerParams(bool enable, unsigned int period)
```
I think this is no longer needed.

mphoward commented on src/PlanarMovingHarmonicBarrierGPU.cu:
The boilerplate in the kernel driver functions has changed a bit, please have a look at a file in HOOMD (e.g., in the MPCD folder) for how to update it.

mphoward commented on src/PlanarMovingHarmonicBarrierGPU.cc:
The way autotuners are used has changed, take a look in HOOMD source code for an example and update as needed.

mphoward commented on src/MovingHarmonicPotential.h:
Do we feel like the cutoff and g parameters are actually useful for us? I don't really use them, so I would be OK if we wanted to remove them in favor of simplicity, going down to just two parameters (spring constant and offset).

kkritikak replied:
I also think it's better to remove them, I also asked Arash and he also said that it's OK  to remove them, currently I haven't removed them but I will remove them soon.

mphoward reviewed (commented):
Thank you Kritika! Here are some comments.

mphoward commented on src/HarmonicBarrier.h:
```diff
@@ -64,14 +67,42 @@ namespace azplugins
+            d["cutoff"] = pybind11::cast(cutoff);
+            return d;
+            }
+        } __attribute__((aligned(16)));
```
After you remove `g` and `cutoff`, this struct will be the same as a `Scalar2`. That means that it should be aligned to 8 bytes for 32-bit floats and 16 bytes for 64-bit floats. You can do this like in the LJ potential:
https://github.com/glotzerlab/hoomd-blue/blob/f0415ec516c84f203d99d96f226759034683221a/hoomd/md/EvaluatorPairLJ.h#L151-L155

mphoward commented on src/HarmonicBarrier.h:
```diff
@@ -82,19 +113,43 @@ class PYBIND11_EXPORT ImplicitEvaporator : public ForceCompute
+        setParams(type_idx, h_params);
+        }
+
+    pybind11::dict getParams(std::string type_name)
```
For completeness, I would like to see this as a Python-specific getter, then also include a C++ getter `getParams` that just returns the actual struct (the complementary method to `setParams`). `getParamsPython` can internally use `getParams` before converting it to a dict.
```suggestion
    pybind11::dict getParamsPython(std::string type_name)
```

mphoward commented on src/HarmonicBarrierGPU.cc:
```diff
@@ -0,0 +1,49 @@
+                                       std::shared_ptr<Variant> interf)
+    : HarmonicBarrier(sysdef, interf)
+    {
+    m_tuner.reset(new Autotuner<1>({AutotunerBase::makeBlockSizeRange(this->m_exec_conf)},
+                                   this->m_exec_conf,
+                                   "harmonic_barrier"));
+    this->m_autotuners.push_back(m_tuner);
```
Since this is not a derived class template, `this->` is not necessary here
```suggestion
    m_tuner.reset(new Autotuner<1>({AutotunerBase::makeBlockSizeRange(m_exec_conf)},
                                  m_exec_conf,
                                   "harmonic_barrier"));
   m_autotuners.push_back(m_tuner);
```

mphoward commented on src/PlanarHarmonicBarrier.cc:
```diff
@@ -3,45 +3,48 @@
+void PlanarHarmonicBarrier::computeForces(uint64_t timestep)
     {
-    ImplicitEvaporator::computeForces(timestep);
+    HarmonicBarrier::computeForces(timestep);
```
I don't think any of the parent methods actually do anything, so there is no need to invoke this here.
```suggestion
```

kkritikak replied:
Thank you for pointing that out! I realized the parent methods actually throw a [warning](https://github.com/mphowardlab/azplugins/blob/v0.11.0/azplugins/ImplicitEvaporator.cc#L41) about the virial contributions, so I thought it might be better to keep it for now. That said, I’m not entirely sure about it—please let me know your thoughts, and I’m happy to remove it if needed!

mphoward replied:
You are correct, I missed that! Please leave these calls in.

mphoward commented on src/PlanarHarmonicBarrierGPU.cc:
```diff
@@ -0,0 +1,87 @@
+ */
+void PlanarHarmonicBarrierGPU::computeForces(uint64_t timestep)
+    {
+    HarmonicBarrierGPU::computeForces(timestep);
```
Same here
```suggestion
```

mphoward commented on src/PlanarHarmonicBarrierGPU.cuh:
```diff
@@ -0,0 +1,37 @@
+    {
+
+//! Kernel driver to evaluate PlanarHarmonicBarrierGPU force
+cudaError_t compute_harmonic_force(Scalar4* d_force,
```
This function and the corresponding kernel should be renamed something more descriptive & specific to the class:
```suggestion
cudaError_t compute_force_planar_harmonic_barrier(Scalar4* d_force,
```

mphoward commented on src/external.py:
```diff
@@ -0,0 +1,61 @@
+from hoomd.data.typeconverter import OnlyTypes, variant_preprocessing
+from hoomd.md import force
+
+class HarmonicBarrier(force.Force):
```
You should make these separate classes that mirror the C++ side (`PlanarHarmonicBarrier` and `SphericalHarmonicBarrier`). They can both derive from a common `HarmonicBarrier` class though! You might follow a pattern like I did here:
https://github.com/mphowardlab/azplugins/blob/fb0ac2fe6cb894e8a437e934f8197897f3d6ad49/src/compute.py#L17

mphoward commented on src/external.py:
```diff
@@ -0,0 +1,61 @@
+
+class HarmonicBarrier(force.Force):
+
+    _ext_module = _azplugins
+#    _cpp_class_name = "PlanarHarmonicBarrier"
```
These are not needed in our own plugin. Specify the module explicitly when you return the class name
```suggestion
```

mphoward reviewed (changes requested):
Looks great! Please proceed with removing those two arguments as well as addressing the mostly small comments below. Last thing we need is to port some of them unit tests that validate these objects.

kkritikak commented:
Thanks for the feedback! 🙌
I’ve addressed most of the comments by removing the cutoff and g parameters and also improved `external.py`.  have also added a new pytest for external.py. Let me know if there’s anything else to adjust. Could you please have a look whenever you have time!

mphoward commented on src/external.py:
```diff
@@ -0,0 +1,188 @@
+
+
+class PlanarHarmonicBarrier(HarmonicBarrier):
+    r"""Applies a purely repulsive harmonic potential to particles
+        near an interface in planar geometry..
```
The first line of a docstring should be a short description.
```suggestion
    r"""Planar harmonic barrier normal to *z*.
```

mphoward commented on src/external.py:
```diff
@@ -0,0 +1,188 @@
+        near an interface in planar geometry..
+
+    Args:
+        interface (`hoomd.variant`) : `z` position of the interface
```
Backticks are used for cross referencing. The type should also be variant_like.
```suggestion
        interface (`hoomd.variant.variant_like`) : *z* position of the interface
```

mphoward commented on src/external.py:
```diff
@@ -0,0 +1,188 @@
+    Args:
+        interface (`hoomd.variant`) : `z` position of the interface
+
+    `PlanarHarmonicBarrier` apply a purely harmonic potential in planar geometry
+    that pushes down on nonvolatile solutes. The potential is truncated at its minimum
+    and have following form:
```
```suggestion
    `PlanarHarmonicBarrier` apply a purely harmonic potential in planar geometry
    with a normal in the :math:`+z` direction:
```

mphoward commented on src/external.py:
```diff
@@ -0,0 +1,188 @@
+
+    .. math::
+
+        \begin{eqnarray*}
+        V(d) = & 0 & d < H \\
+               & \frac{\kappa}{2} (d-H)^2 & H < d
+        \end{eqnarray*}
```
I don't think equation array is needed, but you do want to use cases to make a piecewise function. Also, use the real coordinates now that we have specialized documentation:
```suggestion
        U(z) = \begin{cases}
         0 & z \le H \\
         \dfrac{\kappa}{2} (z-H)^2 & z > H
         \end{cases}
```

mphoward commented on src/external.py:
```diff
@@ -0,0 +1,188 @@
+               & \frac{\kappa}{2} (d-H)^2 & H < d
+        \end{eqnarray*}
+
+    Here, the interface is located at height *H*, and may change with time.
+    The effective interface position *H* may be modified per-particle-type
+    using a *offset* (*offset* is added to *H* to determine the effective *H*)
+    :math:`\kappa` is a spring constant setting the strength of the interface
+    (:math:`\kappa` is a proxy for the surface tension).
```
```suggestion
    Here, the interface is located at height *H*, which may change with time.
    `interface` specifies the nominal *H*, which may then be modified
    per-particle-type by adding an ``offset``. :math:`\kappa` is a spring
    constant setting the strength of the interface, similar to a surface tension.
```

mphoward commented on src/external.py:
```diff
@@ -0,0 +1,188 @@
+    Example::
+
+        interf = hoomd.variant.Ramp(A=50.0, B=25.0, t_start=100, t_ramp=1e6)
+        evap = hoomd.azplugins.external.PlanarHarmonicBarrier(interface = interf)
+
+        evap.params['S'] = dict(k=50.0, offset=0.0)
+        evap.params['B'] = dict(k=50.0*2*2, offset=0.0)
```
```suggestion
        evap = hoomd.azplugins.external.PlanarHarmonicBarrier(interface=interf)
        evap.params['S'] = dict(k=50.0, offset=0.0)
        evap.params['B'] = dict(k=50.0*2**2, offset=0.0)
```
Please also add a short description of what this example is implementing.

mphoward commented on src/external.py:
```diff
@@ -0,0 +1,188 @@
+        evap.params['S'] = dict(k=50.0, offset=0.0)
+        evap.params['B'] = dict(k=50.0*2*2, offset=0.0)
+
+    .. py:attribute:: params
+
+        The `PlanarHarmonicBarrier` parameters. The dictionary has the following
+        keys:
+
+        * ``k`` (`float`, **required**) - Spring constant :math:`k`
+          :math:`[\mathrm{energy} \cdot \mathrm{length}^{-2}]`
+        * ``offset`` (`float`, **required**) - Shift amount per-particle-type :math:`H`
+          :math:`[\mathrm{length}]`
+
+
+        Type: :class:`~hoomd.data.typeparam.TypeParameter` [`tuple`
+        [``particle_type``], `dict`]
```
This should be removed since it is common to the base class.

mphoward commented on src/external.py:
```diff
@@ -0,0 +1,188 @@
+        Type: :class:`~hoomd.data.typeparam.TypeParameter` [`tuple`
+        [``particle_type``], `dict`]
+
+    .. warning::
+        Virial calculation has not been implemented for this model because it is
+        nonequilibrium. A warning will be raised if any calls to
+        :py:class:`hoomd.logging.Logger` are made because the logger always requests
+        the virial flags. However, this warning can be safely ignored if the
+        pressure (tensor) is not being logged or the pressure is not of interest.
```
Move this warning to the base class.

mphoward commented on src/external.py:
```diff
@@ -0,0 +1,188 @@
+
+    """
+
+    def __init__(self, interface=None):
+        super().__init__(interface=interface, geometry="film")
```
You are just going to call the parent constructor after you make the fixes above, so this can just be pass.
```suggestion
    pass
```

mphoward commented on src/external.py:
```diff
@@ -0,0 +1,188 @@
+        super().__init__(interface=interface, geometry="film")
+
+
+class SphericalHarmonicBarrier(HarmonicBarrier):
```
Apply similar changes as for the planar harmonic class here.

mphoward commented on src/pytest/test_external.py:
```diff
@@ -0,0 +1,217 @@
+    sim.run(1)
+
+    # Test forces on particles
+    forces = barrier.forces
+    energies = barrier.energies
```
These attributes are only valid on the root rank in MPI simulations. Please shift these inside a check on `sim.device.communicator.rank`

mphoward commented on src/pytest/test_external.py:
```diff
@@ -0,0 +1,217 @@
+    energies = barrier.energies
+    # particle 0 is outside interaction range
+    assert numpy.isclose(energies[0], 0.0)
+    assert numpy.isclose(forces[0][0], 0.0, atol=1e-4)
+    assert numpy.isclose(forces[0][1], 0.0, atol=1e-4)
+    assert numpy.isclose(forces[0][2], 0.0, atol=1e-4)
```
You can use `assert_allclose` to check multiple values at once, like:
```suggestion
    assert numpy.testing.assert_allclose(forces[0], [0, 0, 0], atol=1e-4)
```

mphoward commented on src/pytest/test_external.py:
```diff
@@ -0,0 +1,217 @@
+    return snap_
+
+
+@pytest.mark.parametrize("cls", [external.SphericalHarmonicBarrier], ids=["Spherical"])
+def test_spherical_harmonic_barrier(
+    simulation_factory, cls, integrator, snap, custom_variant_fixture
+):
+    sim = simulation_factory(snap)
+    sim.operations.integrator = integrator
+
+    barrier = cls(interface=custom_variant_fixture)
```
Since this is a test for a specific class, you don't need to parametrize the test:
```suggestion
def test_spherical_harmonic_barrier(
    simulation_factory, cls, integrator, snap, custom_variant_fixture
):
    sim = simulation_factory(snap)
    sim.operations.integrator = integrator

    barrier = hoomd.azplugins.external.SphericalHarmonicBarrier(interface=custom_variant_fixture)
```

mphoward commented on src/pytest/test_external.py:
```diff
@@ -0,0 +1,217 @@
+# Part of azplugins, released under the BSD 3-Clause License.
+
+import hoomd
+from hoomd.azplugins import external
```
I prefer not to use from imports in unittests because it lets us check that we have internally imported modules as we intend the user to be able to use them.
```suggestion
import hoomd.azplugins
```

mphoward reviewed (changes requested):
Thank you Kritika! I have some points that still need to be addressed, but overall this looks great and is close to merging. Please let me know if you have any questions.

kkritikak commented:
Thanks! I’ve addressed most of the comments. Please let me know if there’s anything else which can be improved.

mphoward commented on src/external.py:
```diff
@@ -0,0 +1,147 @@
+        [``particle_type``], `dict`]
+
+    .. warning::
+        Virial calculation has not been implemented for this model because it is
+        nonequilibrium. A warning will be raised if any calls to
+        :py:class:`hoomd.logging.Logger` are made because the logger always requests
+        the virial flags. However, this warning can be safely ignored if the
+        pressure (tensor) is not being logged or the pressure is not of interest.
```
```suggestion
        Virial calculation has not been implemented for this external field.
```

mphoward commented on src/external.py:
```diff
@@ -0,0 +1,147 @@
+        interf = hoomd.variant.Ramp(A=50.0, B=25.0, t_start=100, t_ramp=1e6)
+        evap = hoomd.azplugins.external.PlanarHarmonicBarrier(interface=interf)
+        # small particle has diameter 1.0
+        evap.params['S'] = dict(k=50.0, offset=0.0)
+        # big particle is twice as large (diameter 2.0), so coefficient is scaled
+        evap.params['B'] = dict(k=50.0*2*2, offset=0.0)
```
```suggestion
        evap.params['S'] = dict(k=50.0, offset=-0.5)
        # big particle is twice as large (diameter 2.0), so coefficients are scaled
        evap.params['B'] = dict(k=50.0*2*2, offset=-1.0)
```

mphoward commented on src/external.py:
```diff
@@ -0,0 +1,147 @@
+    Args:
+        interface (`hoomd.variant.variant_like`) : *z* position of the interface
+
+    `PlanarHarmonicBarrier` apply a purely harmonic potential in planar geometry
+    with a normal in the :math:`+z` direction:
```
```suggestion
    `PlanarHarmonicBarrier` applies a purely harmonic potential in a planar
    geometry with a normal in the :math:`z` direction. Particles are pushed
    in the :math:`-z` direction when they are above the `interface`:
```

mphoward commented on src/external.py:
```diff
@@ -0,0 +1,147 @@
+    Args:
+        interface (`hoomd.variant.variant_like`): position of the interface
+
+    .. py:attribute:: params
```
Interface also needs to be documented as an attribute.

mphoward commented on src/external.py:
```diff
@@ -0,0 +1,147 @@
+    This class should not be instantiated directly. Use a derived type.
+
+    Args:
+        interface (`hoomd.variant.variant_like`): position of the interface
```
```suggestion
        interface (`hoomd.variant.variant_like`): Position of the interface
```

mphoward commented on src/external.py:
```diff
@@ -0,0 +1,147 @@
+
+
+class HarmonicBarrier(force.Force):
+    r"""Applies harmonic barrier.
```
```suggestion
    r"""Repulsive barrier implemented as a harmonic potential.
```

mphoward commented on src/external.py:
```diff
@@ -0,0 +1,147 @@
+        \end{cases}
+
+    Here, the interface is located at height *H*, which may change with time.
+    ``interface`` specifies the nominal *H*, which may then be modified
```
```suggestion
    `interface` specifies the nominal *H*, which may then be modified
```

kkritikak replied:
I wanted to inform you that a warning is being raised during documentation compilation because there is no existing cross-reference for interface. As a result, Sphinx throws the following warning, causing the ReadTheDocs tests to fail:
`hoomd/azplugins/external.py:docstring of hoomd.azplugins.external.PlanarHarmonicBarrier:6: WARNING: 'any' reference target not found: interface [ref.any]`
If I enclose the interface in double backticks (interface) it doesn't cause any problem. Is there a better way to resolve this issue?

mphoward replied:
Oh, yes, that is because that attribute is in the base class, I think, and is not inherited in the documentation. Yes, it is OK to use the double backticks to fix that warning.

mphoward commented on src/external.py:
```diff
@@ -0,0 +1,147 @@
+    r"""Spherical harmonic barrier.
+
+    Args:
+        interface (`hoomd.variant.variant_like`) : Radius of spherical interface.
```
```suggestion
        interface (`hoomd.variant.variant_like`) : Radius of sphere.
```

mphoward commented on src/external.py:
```diff
@@ -0,0 +1,147 @@
+    Args:
+        interface (`hoomd.variant.variant_like`) : Radius of spherical interface.
+
+    `SphericalHarmonicBarrier` applies a purely harmonic potential to particles
+    near a spherical interface that pushes particles inward toward the center
+    of the sphere. The potential is truncated at its minimum
+    and has the following form:
```
```suggestion
    `SphericalHarmonicBarrier` applies a purely harmonic potential to particles
    outside the radius of a sphere, acting to push them inward:
```

mphoward commented on src/external.py:
```diff
@@ -0,0 +1,147 @@
+    of the sphere. The potential is truncated at its minimum
+    and has the following form:
+
+    .. math::
```
Fix the documentation from here to the end to make it mirror the planar geometry documentation. Those changes were not applied here.

mphoward reviewed (changes requested):
Thanks Kritika! There are still some issues with  the  Python documentation. Can you please carefully go through these and make sure you have implemented everything that is needed? Especially keep an eye on the sphere class, since I think the changes for the planar geometry were not propagated there.

mphoward commented on src/PlanarBarrierEvaluator.h:
```diff
@@ -0,0 +1,67 @@
+            return make_scalar4(0, 0, 0, 0);
+            }
+
+        const Scalar f = -k * dy;               // z component of force
```
```suggestion
        const Scalar f = -k * dy;               // y component of force
```

mphoward commented on doc/module-azplugins-external.rst:
```diff
@@ -0,0 +1,27 @@
+.. rubric:: Details
+
+.. automodule:: hoomd.azplugins.external
+    :synopsis: External barrier.
```
```suggestion
    :synopsis: External potentials.
```

mphoward commented on src/SphericalBarrierEvaluator.h:
```diff
@@ -0,0 +1,71 @@
+        }
+
+    private:
+    Scalar m_R; // Radius of sphere
```
```suggestion
    Scalar m_R; //<! Radius of sphere
```

mphoward commented on src/PlanarBarrierEvaluator.h:
```diff
@@ -0,0 +1,67 @@
+        }
+
+    private:
+    Scalar m_H; // y position of interface
```
```suggestion
    Scalar m_H; //<! y position of interface
```

mphoward commented on src/external.py:
```diff
@@ -0,0 +1,154 @@
+# Copyright (c) 2021-2024, Auburn University
+# Part of azplugins, released under the BSD 3-Clause License.
+
+"""External harmonic potentials."""
```
```suggestion
"""External potentials."""
```

mphoward commented on src/external.py:
```diff
@@ -0,0 +1,154 @@
+
+    .. py:attribute:: location
+
+        Location of the barrier. The meaning of this position is interpreted
```
```suggestion
        Location of the barrier. The meaning of this location is interpreted
```

mphoward commented on src/external.py:
```diff
@@ -0,0 +1,154 @@
+        location (`hoomd.variant.variant_like`): *y* position of the
+            barrier.
+
+    `PlanarHarmonicBarrier` applies a purely harmonic potential in a planar
+    geometry with a normal in the :math:`y` direction. Particles are pushed
+    in the :math:`-y` direction when they are above the ``location``:
```
```suggestion
    `PlanarHarmonicBarrier` applies a purely repulsive harmonic potential
    in a planar geometry with a normal in the :math:`y` direction. Particles
    are pushed in the :math:`-y` direction when they are above the
    ``location``:
```

mphoward commented on src/external.py:
```diff
@@ -0,0 +1,154 @@
+    Args:
+        location (`hoomd.variant.variant_like`): Radius of sphere.
+
+    `SphericalHarmonicBarrier` applies a purely harmonic potential to particles
+    outside the radius of a sphere, acting to push them inward:
```
```suggestion
    `SphericalHarmonicBarrier` applies a purely repulsive harmonic potential to
    particles outside the radius of a sphere, acting to push them inward:
```

mphoward reviewed (approved):
Thank you for your work on this! I refactored your code to use a template class to reduce duplication and number of files we have to maintain. I also fixed some issues with the documentation and a deadlock when the unit tests were run with MPI. I will batch commit the small changes I marked here during review, and merge after the unit tests pass. I appreciate you taking care of this!
